materialize-redshift/staged_file.go

```diff
@@ -74,6 +76,16 @@ func (f *stagedFile) start(ctx context.Context) {
 			Body:   r,
 		})
 		if err != nil {
+			// This send cannot block since we are the only sender on the buffered channel. The
+			// error value sent here may be picked up in a call to encodeRow. Or if the error
+			// results after a flush begins, it will retrieved via group.Wait() from the error
+			// returned below.
+			f.uploaderErrCh <- err
+
+			// Close the writer channel now that the uploader has failed and no further writes to
+			// the pipe will be read. This will cause new & outstanding writes to the pipe to return
+			// a "closed pipe" error.
+			f.pipeWriter.Close()
```

Comment:
this can race with a concurrent use of `pipeWriter`. I think you want to be closing `r` here, which also propagates ErrClosedPipe

https://pkg.go.dev/io#PipeReader.Close

Reply:
Also, I think you could use PipeReader.CloseWithError to deliver the causal error to the write loop, rather than having `uploadErrCh` ?